File: 695_Max_Area_of_Island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
      int m = 0, xs = grid[0].size(), ys = grid.size();
  
      for(int x = 0; x < xs ; x++)
        for(int y = 0; y < ys ; y++)                     //search in our array 1 
          if(grid[y][x]){                                //if find out
            queue <pair<int,int>> q;                     //queue for BFS (as visited will use our array: 0 - is visited)
            int tmp = 0;                                 //area for current island 
            q.push({x,y});
            while(!q.empty()){                           //BFS
              int xx = q.front().first, yy = q.front().second;q.pop();
              if(!grid[yy][xx]) continue;                //if visited 
              tmp++;                                     //if not
              grid[yy][xx] = 0;                          //check current cell as visited 
              if(yy > 0 and grid[yy-1][xx]) q.push({xx,yy-1});             //for next level of BFS
              if(yy < ys - 1 and grid[yy+1][xx]) q.push({xx,yy+1});
              if(xx > 0 and grid[yy][xx-1]) q.push({xx-1,yy});
              if(xx < xs - 1 and grid[yy][xx+1]) q.push({xx+1,yy});
            }
            if(tmp > m) m = tmp;                  //check current area and previous maximal area
          } 
      
      return m;    
    }
};

File: 695_Max_Area_of_Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        row, col = len(grid), len(grid[0])
        max_area = 0
        if not grid or not grid[0]:
            return 0

        def dfs(x,y):
            cur_area= 1
            for nx,ny in [[x+1,y],[x-1,y],[x,y+1],[x,y-1]]:
                if 0<=nx<row and 0<=ny<col and grid[nx][ny]==1: 
                    grid[nx][ny]=2
                    cur_area += dfs(nx,ny)

            return cur_area

        for x in range(row):
            for y in range(col):
                if grid[x][y]==1:
                    grid[x][y]=2
                    max_area=max(max_area,dfs(x,y))

        return max_area